### DISCUSSION TITLE
Expand keyword 'default' to support custom default value.

### DISCUSSION THREAD

**Discussion by Qiu233** [dotnet/csharplang#6271] [Language Ideas] [answered]
In C#11, now that we have `static abstract members`, it'd be better to have something like rust's `Default` trait like this:
```rust
#[derive(Default)]
pub struct A {
    pub B: u64,
    pub C: Bytes,
}
```

The version of C# should be like this:
```C#
public interface IHasDefaultValue<TSelf> where TSelf : IHasDefaultValue<TSelf>
{
	static abstract TSelf GetDefaultValue();
}

public class TestClass : IHasDefaultValue<TestClass>
{
	public int V;
	public static TestClass GetDefaultValue()
	{
		return new TestClass() { V = 20 };
	}
}
```
or the property version:
```C#
public interface IHasDefaultValue<TSelf> where TSelf : IHasDefaultValue<TSelf>
{
	static abstract TSelf DefaultValue { get; }
}

public class TestClass : IHasDefaultValue<TestClass>
{
	public int V;

	public static TestClass DefaultValue => new() { V = 20 };
}
```

Then `default(TestClass)` should be resolved to `TestClass.DefaultValue `.

**Comment by svick** [comment, marked as answer]
For classes, the `default` value is currently always `null`. I think it would be too confusing to change that. Especially since it would mean that `default` can now perform an allocation.

**Comment by Richiban** [comment]
While I agree with @svick that, unfortunately, it's not realistic to change the meaning of `default(T)` now, it's still very worthwhile to have a "blessed" `IDefaultable<Self>` interface in the BCL. It's not an issue for this repo though.
